### ISSUE TITLE
Miovision - Light Autos - late night data drops

### ISSUE THREAD

**Issue by aharpalaniTO** [CityofToronto/bdit_data-sources#323]
There appears to be an issue in which data for light autos appears to be truncated for certain intersections and days. Specifically, there appears to be data for cyclists, pedestrians, etc., but light auto volumes are erroneously zero.  While for some intersections this doesn't appear to be a problem at all, for other it's fairly significant (e.g. intersection_uid = 2).

High-level identification of the problem:
```
-- expectation is that if things are normal, this should return next to no results (with possible exception of road closures where peds are still permitted)
WITH ind AS (
SELECT intersection_uid, datetime_bin
FROM miovision_api.volumes_15min
WHERE EXTRACT(HOUR FROM datetime_bin) >= 22
GROUP BY intersection_uid, datetime_bin
HAVING SUM(CASE WHEN classification_uid = 1 THEN volume ELSE NULL END) = 0 AND SUM(volume) > 0 -- where light volumes are 0 in any 15-minute bin, but other modes have volumes
	)
	
	SELECT intersection_uid, intersection_name, date_trunc('month', datetime_bin) AS mth, COUNT(1) as cnt -- maximum value here is ~360 (4 bins/hr * 3 hrs * 30 days)
	FROM ind
	INNER JOIN miovision_api.intersections USING (intersection_uid)
	GROUP BY intersection_uid, intersection_name, date_trunc('month', datetime_bin)
	ORDER BY intersection_uid, date_trunc('month', datetime_bin)
```

Crude confirmation that this issue also exists in `miovision_api.volumes`:
```
-- expectation is that if things are normal, the distribution should be entirely max_hr = 23, with a few scattered exceptions for when cameras go down midday

WITH ind AS (
SELECT intersection_uid, datetime_bin::date AS dt, MAX(EXTRACT(HOUR FROM datetime_bin)) AS max_hr
FROM miovision_api.volumes
WHERE classification_uid = 1
	GROUP BY intersection_uid, datetime_bin::date
)

SELECT intersection_uid, max_hr, COUNT(1)
FROM ind
GROUP BY intersection_uid, max_hr
ORDER BY intersection_uid, max_hr
```

Possible causes:
1, something weird is happening in transform from raw data pull to `miovision_api.volumes`
2. API script is hitting some sort of max row constraint (similar to an issue we had with Blip)
3. Issue on Miovision's end (the API doesn't have all the data available)

**Comment by radumas** [comment]
Should've spotted that earlier
![image](https://user-images.githubusercontent.com/10623872/91313420-5cb02080-e783-11ea-87af-b5074edd6465.png)

**Comment by radumas** [comment]
The problem actually appears to be broader to all classifications in the tmc api call

```sql
SELECT date_trunc('hour', datetime_bin)  as hh, classification, SUM(volume) as volume 
FROM miovision_api.volumes 
NATURAL JOIN miovision_api.classifications
WHERE
datetime_bin >= '2020-08-05 16:00' AND datetime_bin < '2020-08-06 01:00'
AND intersection_uid IN (38) 
-- AND classification_uid NOT IN (2,6)
GROUP BY hh, classification
ORDER BY hh, classification

"2020-08-05 17:00:00";"Light";25
"2020-08-05 17:00:00";"Pedestrian";752
"2020-08-05 17:00:00";"SingleUnitTruck";2
"2020-08-05 17:00:00";"WorkVan";1
"2020-08-05 18:00:00";"Pedestrian";621
"2020-08-05 19:00:00";"Pedestrian";569
"2020-08-05 20:00:00";"Pedestrian";657
"2020-08-05 21:00:00";"Pedestrian";341
"2020-08-05 22:00:00";"Pedestrian";201
"2020-08-05 23:00:00";"Pedestrian";122
"2020-08-06 00:00:00";"ArticulatedTruck";1
```

Testing the api
```python
import requests
import pytz
from datetime import datetime
import json

local_tz=pytz.timezone('US/Eastern') 
url='https://api.miovision.com/intersections/' 
tmc_endpoint = '/tmc'
start_date = datetime(2020,8,20,17)
start_time = local_tz.localize(start_date)
end_time = local_tz.localize(datetime(2020,8,21,1))
intersection_id1='6174c193-13a9-4056-832b-f80783eaf333'
headers={'Content-Type':'application/json','Authorization':key} 
params = {'endTime': end_time, 'startTime' : start_time} 
response=requests.get(url+intersection_id1+tmc_endpoint, params=params, headers=headers)
tmc= json.loads(response.content.decode('utf-8')
tmc
{'class': 'Light',
  'entrance': 'S',
  'exit': 'W',
  'qty': 1,
  'timestamp': '2020-08-20T18:24:00.000-04:00'},
 {'class': 'Bicycle',
  'entrance': 'W',
  'exit': 'E',
  'qty': 3,
  'timestamp': '2020-08-20T18:24:00.000-04:00'},
 {'class': 'Light',
  'entrance': 'W',
  'exit': 'E',
  'qty': 4,
  'timestamp': '2020-08-20T18:24:00.000-04:00'},
 {'class': 'Light',
  'entrance': 'W',
  'exit': 'S',
  'qty': 1,
  'timestamp': '2020-08-20T18:24:00.000-04:00'},
 {'class': 'Light',
  'entrance': 'E',
  'exit': 'W',
  'qty': 1,
  'timestamp': '2020-08-20T18:25:00.000-04:00'}
```

**Comment by aharpalaniTO** [comment]
good catch. forgot that ped and cycling crosswalk volumes come in via a separate API call

**Comment by radumas** [comment]
Trying to repull data for the entire day but running into issues with unique keys 💢 

```bash
/etc/airflow/data_scripts/.venv/bin/python3 /etc/airflow/dev_scripts/volumes/miovision/api/intersection_tmc.py run-api --path /etc/airflow/data_scripts/volumes/miovision/api/config.cfg --start_date 2020-08-20 --end_date 2020-08-21
```

**Comment by radumas** [comment]
Whelp, no change
```sql
SELECT date_trunc('hour', datetime_bin)  as hh, classification, SUM(volume) as volume 
FROM miovision_api.volumes 
NATURAL JOIN miovision_api.classifications
WHERE
datetime_bin >= '2020-08-20 16:00' AND datetime_bin < '2020-08-21 01:00'
AND intersection_uid IN (38) 
-- AND classification_uid NOT IN (2,6)
GROUP BY hh, classification
ORDER BY hh, classification
```

**Comment by radumas** [comment]
easier check to prove data is missing from the api, a pull for all day data for one intersection:

```python
import requests
import pytz
from datetime import datetime
import json

local_tz=pytz.timezone('US/Eastern') 
url='https://api.miovision.com/intersections/' 
tmc_endpoint = '/tmc'
start_time = local_tz.localize(datetime(2020,8,20))
end_time = local_tz.localize(datetime(2020,8,21,1))
intersection_id1='6174c193-13a9-4056-832b-f80783eaf333'
headers={'Content-Type':'application/json','Authorization':key} 
params = {'endTime': end_time, 'startTime' : start_time} 
response=requests.get(url+intersection_id1+tmc_endpoint, params=params, headers=headers)
tmc= json.loads(response.content.decode('utf-8')
missing_data = [record for record in tmc if datetime.strptime(record['timestamp'][:-6], '%Y-%m-%dT%H:%M:%S.%f') >= datetime(2020,8,20,17)]
print(len(missing_data))
# 0 
```

for only data after 1700:
```python
start_date = datetime(2020,8,20,17)
start_time = local_tz.localize(start_date)
end_time = local_tz.localize(datetime(2020,8,21,1))
intersection_id1='6174c193-13a9-4056-832b-f80783eaf333'
headers={'Content-Type':'application/json','Authorization':key} 
params = {'endTime': end_time, 'startTime' : start_time} 
response=requests.get(url+intersection_id1+tmc_endpoint, params=params, headers=headers)
tmc= json.loads(response.content.decode('utf-8')
missing_data = [record for record in tmc if datetime.strptime(record['timestamp'][:-6], '%Y-%m-%dT%H:%M:%S.%f') >= datetime(2020,8,20,17)]
print(len(missing_data))
#4500
```

**Comment by radumas** [comment]
There's a `time_delta` parameter that get defined 

https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/api/intersection_tmc.py#L273

And then used in this loop

https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/api/intersection_tmc.py#L299

To my eyes, unfortunately, at the end here.

https://github.com/CityofToronto/bdit_data-sources/blob/a22e407b09534dc01da99e15893725c2917f0248/volumes/miovision/api/intersection_tmc.py#L342-L345

I think it might be an opportune moment to shift this to what I think would be a [more maintainable solution](https://stackoverflow.com/a/1061779/4047679)

Oh also processing should move out and still be on a daily basis.

**Comment by cczhu** [comment]
Quite annoyingly, according to the `rrule` documentation the frequency must follow:

```
Where freq must be one of YEARLY, MONTHLY, WEEKLY, DAILY, HOURLY, MINUTELY,
or SECONDLY.
```

I think the [winning post](https://stackoverflow.com/a/1060330) in the same answer offers a better solution

```python
from datetime import timedelta, date

def daterange(start_date, end_date):
    for n in range(int((end_date - start_date).days)):
        yield start_date + timedelta(n)

start_date = date(2013, 1, 1)
end_date = date(2015, 6, 2)
for single_date in daterange(start_date, end_date):
    print(single_date.strftime("%Y-%m-%d"))
```

**Comment by aharpalaniTO** [comment]
@cczhu - as promised, the other high-level check for missing data

SQL (takes ~5 minutes to run)
```
-- checks for proportion of days that don't end in 23:59 (i.e. days where the last data for light autos is before the end of the day)
-- 0.0% isn't necessarily achievable if cameras have legitimately gone down, but should be a good enough comparative

WITH X AS (
	SELECT intersection_uid, datetime_bin::date, EXTRACT(EPOCH FROM ('23:59'::time without time zone - MAX(datetime_bin::time without time zone)))/60/(60*24)  AS time_missing
	FROM miovision_api.volumes
	WHERE classification_uid = 1
	GROUP BY intersection_uid, datetime_bin::date
)

SELECT date_trunc('month', datetime_bin)::date AS mth, AVG(time_missing) AS pct_missing
FROM X
GROUP BY date_trunc('month', datetime_bin)
ORDER BY date_trunc('month', datetime_bin)
```

Result Set:
mth | pct_missing
-- | --
01/01/2019 | 1.3%
02/01/2019 | 1.2%
03/01/2019 | 2.4%
04/01/2019 | 2.0%
05/01/2019 | 2.3%
06/01/2019 | 3.8%
07/01/2019 | 4.9%
08/01/2019 | 5.4%
09/01/2019 | 5.4%
10/01/2019 | 3.6%
11/01/2019 | 3.2%
12/01/2019 | 1.8%
01/01/2020 | 1.9%
02/01/2020 | 1.8%
03/01/2020 | 1.4%
04/01/2020 | 0.3%
05/01/2020 | 0.5%
06/01/2020 | 2.7%
07/01/2020 | 6.0%
08/01/2020 | 6.1%
09/01/2020 | 6.5%

**Comment by cczhu** [comment]
To test the largest time interval to use when polling the Miovision API without returning large blocks of missing data, I modified Raph's code [above](https://github.com/CityofToronto/bdit_data-sources/issues/323#issuecomment-681085745) to retrieve data from between 2020-08-20 00:00:00 and 2020-08-24 00:00:00 in increments of 4, 6, 12, 24 and 48 hours. Here's a plot of data availability, where grey blocks indicate available data.

![image](https://user-images.githubusercontent.com/442891/93260973-1e00fb00-f770-11ea-9db7-ac03beeea85f.png)

With the exception of a few moments (almost always in the middle of the night) without any vehicles counted, the data is complete when using 4, 6, and 12 hour increments. When switching to 24 hours, only the first ~950 points are available; this is also true for 48 hours.

We're therefore safe in using a 6-hour interval when polling the API. We could probably go up to 12 hours without any loss in data, but should stick to 6 to be safe.

**Comment by cczhu** [comment]
Made a few changes to intersection_tmc.py. Testing this on Adelaide/Spadina (`uid = 2, c3ebb880-a4ed-47ab-8079-a3e29fb1913e`) from 2019-07-10 00:00:00 to 2019-07-12 00:00:00.

On 2020-09-15, the command

```
SELECT COUNT(*)
FROM miovision_api.volumes
WHERE intersection_uid = 2 AND classification_uid != 6 AND datetime_bin BETWEEN '2019-07-10 00:00:00' AND '2019-07-12 00:00:00'
```

returned 19387

```
SELECT COUNT(DISTINCT datetime_bin)
FROM miovision_api.volumes
WHERE intersection_uid = 2 AND classification_uid != 6 AND datetime_bin BETWEEN '2019-07-10 00:00:00' AND '2019-07-12 00:00:00'
```

returned 2589.

Ran

```
python intersection_tmc.py run-api --path /etc/airflow/data_scripts/volumes/miovision/api/config.cfg --start_date 2019-07-10 --end_date 2019-07-12 --intersection 2
```

Following this, the scripts above give 22121 and 2866. (Including 2019-07-12 00:00:00, there are 2881 distinct minute timestamps over 2 days.)

The trouble arises when we try to corroborate these values with ones extracted directly from the API.

**Comment by cczhu** [comment]
Creating a small script to directly extract non-pedestrian counts from the API:

```python
import psycopg2
import math
import pytz
import datetime
import dateutil
import requests
import json
import pandas as pd
import pathlib
import configparser

# Grab Postgres and Miovision credentials.
config = configparser.ConfigParser()
config.read(pathlib.Path.home().joinpath('.charlesconfig').as_posix())
postgres_settings = config['POSTGRES']
miov_token = config['MIOVISION']['key']

local_tz = pytz.timezone('US/Eastern')

url = 'https://api.miovision.com/intersections/' 
tmc_endpoint = '/tmc'

# Extract all non-pedestrian values using the Miovision API from 2020-07-10
# to 2020-07-12 (inclusive) with a cadence of 6 hours.
start_time = local_tz.localize(datetime.datetime(2019, 7, 10))
end_time = local_tz.localize(datetime.datetime(2019, 7, 12, 0, 1, 0))
dt = datetime.timedelta(hours=6)
# Equivalent to intersection_uid = 2
intersection_id1 = "c3ebb880-a4ed-47ab-8079-a3e29fb1913e"
headers={'Content-Type':'application/json', 'Authorization': miov_token}

# Generator for cycling through 6-hour time blocks.
def timerange(start_time, end_time, dt):
    for i in range(math.ceil((end_time - start_time) / dt)):
        c_start = start_time + i * dt
        c_end = min(c_start + dt, end_time)
        yield (c_start, c_end)

tmc_all = []

for (c_start_t, c_end_t) in timerange(start_time, end_time, dt):
    params = {'endTime': c_end_t, 'startTime' : c_start_t} 
    response=requests.get(url + intersection_id1 + tmc_endpoint,
                          params=params, headers=headers)
    tmc = json.loads(response.content.decode('utf-8'))
    tmc_all += tmc
```

retrieves raw data from the API for the same time range and intersection as in the previous comment.

```
len(tmc_all)
```

gives 20266, which is significantly less than 22121.

We can turn `tmc_all` into a Pandas DataFrame:

```python
# Functions from intersections_tmc.py for encoding movement and class.
def get_movement(entrance, exit_dir):
    if (entrance == 'N' and exit_dir =='S'):
        return '1'
    elif entrance == 'S' and exit_dir =='N':
        return '1'
    elif entrance == 'W' and exit_dir =='E':
        return '1'
    elif entrance == 'E' and exit_dir =='W':
        return '1'
    elif entrance == 'S' and exit_dir =='W':
        return '2'
    elif entrance == 'N' and exit_dir =='E':
        return '2'
    elif entrance == 'W' and exit_dir =='N':
        return '2'
    elif entrance == 'E' and exit_dir =='S':
        return '2'
    elif entrance == 'S' and exit_dir =='E':
        return '3'
    elif entrance == 'E' and exit_dir =='N':
        return '3'
    elif entrance == 'N' and exit_dir =='W':
        return '3'
    elif entrance == 'W' and exit_dir =='S':
        return '3'
    return '4'

def apply_get_movement(row):
    return get_movement(row["entrance"], row["exit"])

def get_classification(veh_class):
    if (veh_class == 'Pedestrian'):
        return '6'
    elif veh_class == 'Light':
        return '1'
    elif veh_class == 'Bicycle':
        return '2'
    elif veh_class == 'Bus':
        return '3'
    elif veh_class == 'SingleUnitTruck':
        return '4'
    elif veh_class == 'ArticulatedTruck':
        return '5'
    elif veh_class == 'WorkVan':
        return '8'
    return '0'

# Convert to Pandas DataFrame.
df_tmc = pd.DataFrame(tmc_all)
df_tmc["movement_uid"] = df_tmc.apply(apply_get_movement, axis=1).astype(int)
df_tmc["classification_uid"] = df_tmc["class"].apply(get_classification).astype(int)
df_tmc['datetime_bin'] = pd.to_datetime(df_tmc['timestamp']).dt.tz_convert(local_tz)
df_tmc.rename(columns={'qty': "volume"}, inplace=True)
df_tmc['in_api'] = True
```

```python
df_tmc['datetime_bin'].unique().shape
```

yields 2866. At least this is the same.

**Comment by cczhu** [comment]
We can join the dataframe derived from the API calls with one with the data from Postgres, to determine which data is shared between the two and which is unique to either data source:

```python
# Retrieve data from the Postgres raw data.
sql_query = """SELECT *
FROM miovision_api.volumes
WHERE intersection_uid = 2 AND classification_uid != 6 AND datetime_bin BETWEEN '2019-07-10' AND '2019-07-12'
"""

with psycopg2.connect(database='bigdata', **postgres_settings) as db_con:
    df_pg = pd.read_sql(sql_query, db_con)
    df_pg['datetime_bin'] = df_pg['datetime_bin'].dt.tz_localize(local_tz)
    df_pg['in_postgres'] = True

# Merge all unique
df_combined = pd.merge(
    df_pg[['datetime_bin', 'classification_uid', 'movement_uid', 'volume', 'in_postgres']],
    df_tmc[['datetime_bin', 'classification_uid', 'movement_uid', 'volume', 'in_api']],
    how="outer",
    on=['datetime_bin', 'classification_uid', 'movement_uid', 'volume'])
```

Unfortunately `df_combined` has NaNs in both `in_postgres` and `in_api`, meaning that there is data unique to each data source. This is corroborated by plotting the unique volume counts:

```
%matplotlib notebook
import matplotlib.pyplot as plt

fig, ax = plt.subplots(figsize=(8, 6))

ax.plot(df_combined.loc[df_combined['in_postgres'].isna(), 'datetime_bin'],
        df_combined.loc[df_combined['in_postgres'].isna(), 'volume'], ms=2, marker="o", ls="none",
        label="Only in Postgres")
ax.plot(df_combined.loc[df_combined['in_api'].isna(), 'datetime_bin'],
        df_combined.loc[df_combined['in_api'].isna(), 'volume'], ms=2, mew=5, marker="+", ls="none",
        label="Only from API")
ax.set_xlabel("time")
ax.set_ylabel("volume")
ax.legend(loc=2);
```

![image](https://user-images.githubusercontent.com/442891/93494794-6be64200-f8db-11ea-87cf-02625dcbf143.png)

Curiously there appears to be some correlation between the two datasets.

Looking into classification:

![image](https://user-images.githubusercontent.com/442891/93495048-b2d43780-f8db-11ea-8582-4feb8fee04d8.png)

we notice that there's nothing special about the vehicle class of the data that's unique to the API, but there is for the data that's unique to Postgres, which is the only data that contains class 8, "WorkVan".

Raph found an e-mail from Miovision:

> I believe it may actually be prudent for us in future to be a bit more transparent and surface 'Motorized Vehicles' explicitly, to allow you to do what you want with the data. Will make sure to tell you if/when this gets roadmapped. In the future we can look at explicitly modelling a Streetcar class.

> What Changes will be made to the classes?
> 1. The API will consolidate the classes so that you won't need to do the adding yourself.
> 2. WorkVans should be consolidated into Lights
> 3. Motorized Vehicles will be separated from Buses

If we do `df_tmc['class'].unique()`, we can also find "MotorizedVehicle" as a class, which gets mapped to 0.

So known issues not addressed in our data:
- MotorizedVehicle is a new class separate from Buses.
- WorkVans are now Lights

**Comment by radumas** [comment]
I guess the subtlety to point out is that if some classifications got subsumed into others, the number of rows might not quite match and you should instead be looking at total volume by classification.

By the way you can determine which rows exactly were inserted Tuesday by looking at the `volume_uid` ranges

```sql
SELECT DISTINCT ON (volume_uid/ 100000000) volume_uid, intersection_uid, datetime_bin, classification_uid, leg, movement_uid, volume, volume_15min_tmc_uid
	FROM  miovision_api.volumes_2019
	 WHERE intersection_uid = 2 AND classification_uid != 6 AND datetime_bin BETWEEN '2019-07-10 00:00:00' AND '2019-07-12 00:00:00'
	 ORDER BY (volume_uid/ 100000000)
```

**Comment by cczhu** [comment]
Assuming that rows with `volume_uid = 773728979` or above were newly inserted on Tuesday, 2734 rows were inserted on that day. What confuses me is that if we successfully inserted all rows from the API into the database, why are there any rows that are in the API but not in the database?

While working on this, discovered that I had neglected to include "leg" in the join, thinking that `movement_uid` would include it. This required a rewrite of the script's processing and final merge to:

```python
# Convert to Pandas DataFrame.
df_tmc = pd.DataFrame(tmc_all)
df_tmc["movement_uid"] = df_tmc.apply(apply_get_movement, axis=1).astype(int)
df_tmc["classification_uid"] = df_tmc["class"].apply(get_classification).astype(int)
df_tmc['datetime_bin'] = pd.to_datetime(df_tmc['timestamp']).dt.tz_convert(local_tz)
df_tmc.rename(columns={'qty': "volume"}, inplace=True)
df_tmc["leg"] = df_tmc["entrance"]
df_tmc['in_api'] = True

# Retrieve data from the Postgres raw data.
sql_query = """SELECT *
FROM miovision_api.volumes
WHERE intersection_uid = 2 AND classification_uid != 6 AND datetime_bin BETWEEN '2019-07-10' AND '2019-07-12'
"""

with psycopg2.connect(database='bigdata', **postgres_settings) as db_con:
    df_pg = pd.read_sql(sql_query, db_con)
    df_pg['datetime_bin'] = df_pg['datetime_bin'].dt.tz_localize(local_tz)
    df_pg['in_postgres'] = True
    df_pg['new_insert'] = df_pg['volume_uid'] >= 700000000

# Merge all unique
df_combined = pd.merge(
    df_pg[['datetime_bin', 'classification_uid', 'leg', 'movement_uid',
           'volume', 'in_postgres', 'volume_uid', 'new_insert']],
    df_tmc[['datetime_bin', 'classification_uid', 'leg', 'movement_uid', 'volume', 'in_api']],
    how="outer",
    on=['datetime_bin', 'classification_uid', 'leg', 'movement_uid', 'volume'])
df_combined.sort_values(['datetime_bin', 'classification_uid', 'leg', 'movement_uid', 'volume'], inplace=True)
```

This also includes a `new_insert` column which checks whether the data was inserted on Tuesday by checking if its volume UID is greater than 700,000,000.

Revised diagnostics:

![image](https://user-images.githubusercontent.com/442891/93526832-0491b780-f906-11ea-88bb-01a65d6a7f2c.png)

![image](https://user-images.githubusercontent.com/442891/93526862-11aea680-f906-11ea-9ebe-5a822710088a.png)

Now, looking at the `new_insert` column:

![image](https://user-images.githubusercontent.com/442891/93527230-8aadfe00-f906-11ea-996d-6bb4a9cfea20.png)

So everything that was inserted on Tuesday is included both in the API and Postgres. All data in Postgres but not the API were inserted a prior to Tuesday.

**Comment by cczhu** [comment]
Raph points out that:

- The `miovision.volumes` partition tables (eg. `movision.volumes2019`) have the constraint `ADD CONSTRAINT volumes_2019_intersection_uid_datetime_bin_classification_u_key UNIQUE (intersection_uid, datetime_bin, classification_uid, leg, movement_uid);` (which checks everything except the actual volume).
- We have a new `MotorizedVehicle` class separated out from buses starting November 7, 2019.
- Miovision moved class 8, `WorkVan`, into light vehicles starting at the same time.
- There may have been minor changes to the volumes.

This means that when we update historical data from before November, like we're currently doing, we run into the following:

- Data where the volumes have been changed but nothing else are not inserted into the table. This means that if the light vehicle or bus volumes have changed (because of the issues mentioned above) they would not be updated. This explains why we have rows in the API but not in Postgres, even though 2734 new rows *were* inserted during Tuesday's update: my Pandas join above joins on all available rows, including volume.
- Likewise, we have rows in Postgres that aren't in the API reflecting the older volumes.
- Finally, we have rows that were *successfully* added to Postgres from the API. These come from the new `MotorizedVehicle` class, which is why all class 0 rows were successfully inserted.

This suggests a quick test.

The number of work van and light vehicle detections in the Postgres-exclusive data should roughly equal the number of light vehicle detections in the API-exclusive data.

![image](https://user-images.githubusercontent.com/442891/93605791-33a33a00-f995-11ea-915d-dfa3da11b322.png)

Eh, that's about right.

Also, the number of buses in the Postgres-exclusive data should equal the number of buses in the API-exclusive data plus all the class 0 detections that were recently uploaded.

![image](https://user-images.githubusercontent.com/442891/93605859-4e75ae80-f995-11ea-9e2a-81bca12cefce.png)

🤷‍♀️

Part of the problem is there are too many moving parts to this comparison. Given we now have #329, let's pick a 2 day timespan in late November to do the same test.

**Comment by cczhu** [comment]
From some e-mail sleuthing from Aakash and some more digging from Raph and myself, the switchover of streetcars from buses to generic motorized vehicles and the removal of work vans happened in summer 2020, which makes it difficult to do exact comparisons between data on the API and data on Postgres following a refresh with `intersection_tmc.py` without deleting the data in Postgres first.

One of the few places it's possible to do a check without deletion is Danforth / Woodbine (`uid = 38`, `id = 6174c193-13a9-4056-832b-f80783eaf333`), which Raph discovered (above) has missing data in late August. Ran a data refresh for those days (after casting `classification_uid = 0` to `9`). Following the refresh, data from the API and from Postgres agree with the exception of a single line in each:

![image](https://user-images.githubusercontent.com/442891/93797946-e203e500-fc0a-11ea-8e8c-30603b93371e.png)

This looks like a data correction on the part of Miovision. For details, see `Development Notebooks/02 - Test New Data Insertion.ipynb`.

Conclusions:
- The strong agreement between the API and refreshed data means `intersection_tmc.py` is working in general.
- Minor changes to pre-existing data suggest we either have to dump all existing data or allow for overwrites of existing data with new changes during the refresh.

**Comment by cczhu** [comment]
The issue of entire evening blocks of auto data missing has now been solved by #332, so closing (not sure why this wasn't auto-closed).
